examples/factorial.rs

```diff
@@ -0,0 +1,180 @@
+use std::hash::Hash;
+
+use chiquito::{
+    field::Field,
+    frontend::dsl::circuit, // main function for constructing an AST circuit
+    plonkish::backend::halo2::{chiquito2Halo2, ChiquitoHalo2Circuit}, /* compiles to
+                             * Chiquito Halo2
+                             * backend,
+                             * which can be
+                             * integrated into
+                             * Halo2
+                             * circuit */
+    plonkish::compiler::{
+        cell_manager::SingleRowCellManager, // input for constructing the compiler
+        compile,                            // input for constructing the compiler
+        config,
+        step_selector::SimpleStepSelectorBuilder,
+    },
+    plonkish::ir::{assignments::AssignmentGenerator, Circuit}, // compiled circuit type
+    poly::ToField,
+};
+use halo2_proofs::{dev::MockProver, halo2curves::bn256::Fr};
+
+const MAX_FACTORIAL: usize = 10;
+
+fn generate<F: Field + From<u64> + Hash>() -> (Circuit<F>, Option<AssignmentGenerator<F, u32>>) {
+    //
+    // table for the circuit:
+    // |    step_type      |  i  |  x   |
+    // ----------------------------------
+    // |  setup_step       |  0  |  1   |
+    // |  operation_step   |  1  |  1   |
+    // |  operation_step   |  2  |  2   |
+    // |  operation_step   |  3  |  6   |
+    // |  result_step      |  4  |  24  |
+    // |  result_step      |  4  |  24  |
+    // |  result_step      |  4  |  24  |
+    // ...
+
+    use chiquito::frontend::dsl::cb::*; // functions for constraint building
+
+    let factorial_circuit = circuit::<F, u32, _>("factorial", |ctx| {
+        let i = ctx.shared("i");
+        let x = ctx.forward("x");
+
+        // setup step will make sure the circuit is initialized correctly
+        let setup_step = ctx.step_type_def("setup_step", |ctx| {
```

Comment:
The use of the word setup can be confusing

Reply:
I see the problem. Updated to `FirstStep` now :+1: